agent/consul/state/catalog.go

```diff
@@ -2546,125 +2546,109 @@ func (s *Store) updateTerminatingGatewayNamespace(tx *memdb.Txn, gateway structs
 			continue
 		}
 
-		existing, err := s.serviceTerminatingGateway(tx, sn.ServiceName, &sn.EnterpriseMeta)
+		existing, err := tx.First(gatewayServicesTableName, "id", service.Gateway, sn.CompoundServiceName())
 		if err != nil {
 			return fmt.Errorf("gateway service lookup failed: %s", err)
 		}
-
-		if gs, ok := existing.(*structs.GatewayService); ok && gs != nil {
-			// Return an error if the wildcard is attempting to cover a service specified by a different gateway's config entry
-			if !gs.Gateway.Matches(&gateway) {
-				return fmt.Errorf("service %q is associated with different gateway, %q", gs.Service.String(), gs.Gateway.String())
-			}
-
+		if existing != nil {
 			// If there's an existing service associated with this gateway then we skip it.
 			// This means the service was specified on its own, and the service entry overrides the wildcard entry.
 			continue
 		}
 
-		mapping := &structs.GatewayService{
-			Gateway:     gateway,
-			Service:     structs.NewServiceID(sn.ServiceName, &service.EnterpriseMeta),
-			GatewayKind: structs.ServiceKindTerminatingGateway,
-			KeyFile:     service.KeyFile,
-			CertFile:    service.CertFile,
-			CAFile:      service.CAFile,
-		}
-		if err := tx.Insert(terminatingGatewayServicesTableName, mapping); err != nil {
-			return fmt.Errorf("failed inserting gateway service mapping: %s", err)
+		mapping := service.Clone()
+		mapping.Gateway = gateway
```

Comment:
I don't think you need this update. From what I'm seeing, `ingressConfigGatewayServices` uses the same `gatewayID` to build `service` as the one that gets passed into this function.

Reply:
You're right, pushed an update.